### ISSUE TITLE
Compile libvips with libheif on amazon linux

### ISSUE THREAD

**Issue by estepnv** [libvips/libvips#1844]
```log
/bin/sh ../libtool  --tag=CC   --mode=link clang  -g -O2 -lgirepository-1.0 -lgobject-2.0 -lglib-2.0   -o introspect introspect.o -lgirepository-1.0 -lgobject-2.0 -lglib-2.0   -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS    -pthread -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include              -I/usr/include/libpng15                    -I/usr/local/include   libvips.la -lz   -L/usr/local/lib -lheif     -lpng15    -ltiff   -ljpeg     -Wl,--export-dynamic -pthread -lgmodule-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0   -lexpat       -llcms2               -lm  
libtool: link: clang -g -O2 -o .libs/introspect introspect.o -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -pthread -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/libpng15 -I/usr/local/include -Wl,--export-dynamic -pthread  -lgirepository-1.0 ./.libs/libvips.so -L/usr/local/lib -lz /usr/local/lib/libheif.so -lx265 -lpthread -lpng15 -ltiff -ljpeg -lgmodule-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lexpat -llcms2 -lm -pthread -Wl,-rpath -Wl,/usr/lib64/lib -Wl,-rpath -Wl,/usr/local/lib
./.libs/libvips.so: undefined reference to `vips__heif_suffs'
./.libs/libvips.so: undefined reference to `vips__heif_error'
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [introspect] Error 1
make[2]: Leaving directory `/build/libvips/libvips'
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory `/build/libvips/libvips'
make: *** [install-recursive] Error 1
```
can pls anybody help me with this?
I'm trying to build libvips 8.10.1 on amazon linux
with libheif 1.9.1

here's my Dockerfile
```docker
FROM lambci/lambda:build-ruby2.7

WORKDIR /build

ENV WORKDIR="/build"
ENV INSTALLDIR="/opt"
ENV VIPS_VERSION="8.10.1"

RUN curl -L https://github.com/strukturag/libheif/releases/download/v1.9.1/libheif-1.9.1.tar.gz | tar xz
RUN git clone git://github.com/libvips/libvips.git

RUN yum install -y gtk-doc
RUN yum install -y gobject-introspection \
  gobject-introspection-devel \
  expat-devel \
  libjpeg-turbo-devel \
  libtiff-devel \
  libpng-devel \
  lcms2-devel \
  mercurial

RUN curl http://mirror.centos.org/centos/8/PowerTools/x86_64/os/Packages/nasm-2.13.03-2.el8.x86_64.rpm --output nasm.rpm && yum install -y nasm.rpm

RUN hg clone http://hg.videolan.org/x265
RUN cd ./x265/build/linux && \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig cmake -G "Unix Makefiles" ../../source && \
  make -j6 && \
  make install && \
  ldconfig

RUN cd libheif-1.9.1 && \
  ./autogen.sh && \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig ./configure && \
  make && \
  make install

RUN PKG_CONFIG_PATH=/usr/local/lib/pkgconfig  pkg-config libheif --modversion && exit 1
RUN cd libvips && git checkout "v8.10.1"
RUN cd ./libvips && \
  CC=clang CXX=clang++ \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig \
  ./autogen.sh \
  --prefix=/usr/lib64 \
  --with-jpeg-includes=/usr/include \
  --with-jpeg-libraries=/usr/lib64 \
  --with-png-includes=/usr/include \
  --with-png-libraries=/usr/lib64 \
  --with-tiff-includes=/usr/include \
  --with-tiff-libraries=/usr/lib64 \
  --with-lcms2-includes=/usr/include \
  --with-lcms2-libraries=/usr/lib64 \
  --with-heif-includes=/usr/local/include \
  --with-heif-libraries=/usr/local/lib \
  --disable-static && \
  make install && \
  echo /opt/lib > /etc/ld.so.conf.d/libvips.conf && \
  echo /usr/lib64 > /etc/ld.so.conf.d/libvips.conf && \
  ldconfig

# Copy only needed so files to new share/lib.
#

RUN mkdir -p share/lib && \
  cp -a /usr/lib64/lib/libvips.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libpng*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libjbig*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libblkid*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libuuid*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/liblcms2*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libmount*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libgio*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libjpeg*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/local/lib/libheif*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libtiff*.so* $WORKDIR/share/lib/

# Create sym links for ruby-ffi gem's `glib_libname` and `gobject_libname` to work.
RUN cd ./share/lib/ && \
  ln -s /usr/lib64/libglib-2.0.so.0 libglib-2.0.so && \
  ln -s /usr/lib64/libglib-2.0.so.0 glib-2.0.so.0 && \
  ln -s /usr/lib64/libgobject-2.0.so.0 libgobject-2.0.so && \
  ln -s /usr/lib64/libgobject-2.0.so.0 gobject-2.0.so.0 && \
  ln -s /usr/lib64/libgmodule-2.0.so.0 libgmodule-2.0.so.0 && \
  ln -s /usr/lib64/libexpat.so.1 libexpat.so.1


RUN cd ./share && \
  zip --symlinks -r libvips.zip .

```

**Comment by jaume-pinyol** [comment]
I build it like this in amazon linux2, if it's from any help:
```bash

function assert_lib_installed() {
  lib=$1
  if [ $(pkg-config --list-all | grep "$lib" | wc -l) -ne 1 ]; then
    echo "$lib has not been properly installed"
    exit 1
  else
    echo "$lib has been properly build and installed"
  fi
}

function install_libheif() {
  nasm_version=2.15.05
  echo "Installing nasm version ${nasm_version}"
  curl -L -O https://www.nasm.us/pub/nasm/releasebuilds/${nasm_version}/nasm-${nasm_version}.tar.bz2
  tar xjf nasm-${nasm_version}.tar.bz2
  cd nasm-${nasm_version}
  ./autogen.sh
  ./configure --prefix="/usr"
  make
  sudo make install
  sudo ldconfig
  cd ..

  libde265_version=1.0.7
  echo "Installing libde265 version ${libde265_version}"
  wget https://github.com/strukturag/libde265/releases/download/v${libde265_version}/libde265-${libde265_version}.tar.gz
  tar -xvf libde265-${libde265_version}.tar.gz
  cd libde265-${libde265_version}
  ./autogen.sh
  ./configure --disable-dec265 --disable-sherlock265 --enable-log-info --prefix="/usr"
  make
  sudo make install
  sudo ldconfig
  cd ..
  assert_lib_installed "libde265"

  libx265_version=3.4
  echo "Installing libx265 version ${libx265_version}"
  wget https://github.com/videolan/x265/archive/${libx265_version}.tar.gz
  tar -xvf ${libx265_version}.tar.gz
  cd  x265-${libx265_version}
  cd build/linux
  cmake -G "Unix Makefiles" -DCMAKE_INSTALL_PREFIX="/usr"  ../../source
  make
  sudo make install
  sudo ldconfig
  cd ../../..
  assert_lib_installed "x265"

  echo "Installing libheif version ${LIBHEIF_VERSION}"
  wget https://github.com/strukturag/libheif/releases/download/v${LIBHEIF_VERSION}/libheif-${LIBHEIF_VERSION}.tar.gz
  tar xvf libheif-${LIBHEIF_VERSION}.tar.gz
  cd libheif-${LIBHEIF_VERSION}
  ./autogen.sh
  CPPFLAGS="-I${LIBJPEG_HOME}/include" LIBS="-ldl" ./configure --prefix="/usr"
  make
  sudo make install
  sudo ldconfig
  cd ..
  assert_lib_installed "libheif"

  if [ "$(pkg-config libheif --variable builtin_h265_decoder)" != "yes" ]; then
    echo "Libheif has been build without heif decoder support"
    exit 1
  fi

   if [ "$(pkg-config libheif --variable builtin_h265_encoder)" != "yes" ]; then
    echo "Libheif has been build without heif encoder support"
    exit 1
  fi
}
```
I install it in `/usr` so I don't need to be modifying the `PKG_CONFIG_PATH`

and libvips with this is enough for me:
```
function install_libvips_from_source() {
  echo "Installing Libvips version=${LIBVIPS_VERSION}"
  wget ${VIPS_URL}/v${LIBVIPS_VERSION}/vips-${LIBVIPS_VERSION}.tar.gz -O vips.tar.gz
  tar xzf vips.tar.gz
  cd vips-${LIBVIPS_VERSION}

  CFLAGS=-O3 CXXFLAGS=-O3 ./configure --disable-debug --disable-dependency-tracking \
   --enable-gtk-doc-html=no --enable-gtk-doc=no --enable-pyvips8=no \
   --without-python --prefix=$VIPSHOME

  make V=0
  sudo make install
  sudo ldconfig
  cd ..
  echo "Installed libvips ${LIBVIPS_VERSION}"
}
```

**Comment by jcupitt** [comment]
Hello @estepnv,

This was a bug -- some definitions shared between heifload and heifsave were inside the #ifdef for load, so they were not defined if you were only saving them.

I've fixed it (I think) and the fix should be in 8.10.2. Thanks for reporting this!

**Comment by jcupitt** [comment]
I would build libvips like this:

```
ARG VIPS_VERSION=8.10.1
ARG VIPS_URL=https://github.com/libvips/libvips/releases/download

# selection of packages for libvips -- you might want to expand this
RUN yum install -y \
        expat-devel \
        glib2-devel \
        lcms2-devel \
        libexif-devel \
        libgsf-devel \
        libjpeg-turbo-devel \
        libpng-devel \
        libtiff-devel \
        orc-devel 
  
RUN wget $VIPS_URL/v$VIPS_VERSION/vips-$VIPS_VERSION.tar.gz \
        && tar xzf vips-$VIPS_VERSION.tar.gz \
        && cd vips-$VIPS_VERSION \
        && ./configure \
        && make V=0 \
        && make install
```

1. Don't clone the repo (use the tarball),
2. Don't use `--with-jpeg-includes` etc., autodetect is safer
3. Use orc (you get a nice speedup for things like gaussblur and resize),
4. Use libexif (for metadata handling)

If you want to install to `/opt`, I would build the whole stack to that prefix.

You shouldn't need the symlinks for ruby-vips.

**Comment by jcupitt** [comment]
This dockerfile seems to work for me:

```
FROM lambci/lambda:build-ruby2.7

# general build stuff
RUN yum update -y \
        && yum groupinstall -y "Development Tools" \
        && yum install -y \
                wget \
                mercurial

WORKDIR /build

ENV WORKDIR="/build"
ENV INSTALLDIR="/opt"
ENV PKG_CONFIG_PATH=/usr/local/lib/pkgconfig

RUN curl http://mirror.centos.org/centos/8/PowerTools/x86_64/os/Packages/nasm-2.13.03-2.el8.x86_64.rpm --output nasm.rpm && yum install -y nasm.rpm

RUN hg clone http://hg.videolan.org/x265
RUN cd ./x265/build/linux \
        && cmake -G "Unix Makefiles" ../../source \
        && make -j6 \
        && make install \
        && ldconfig

ARG HEIF_URL=https://github.com/strukturag/libheif/releases/download
ARG HEIF_VERSION=1.9.1

RUN curl -L $HEIF_URL/v$HEIF_VERSION/libheif-$HEIF_VERSION.tar.gz | tar xz \
        && cd libheif-$HEIF_VERSION \
        && ./autogen.sh \
        && ./configure \
        && make V=0 \
        && make install

ARG VIPS_VERSION=8.10.1
ARG VIPS_URL=https://github.com/libvips/libvips/releases/download

# selection of packages for libvips -- you might want to expand this
RUN yum install -y \
        expat-devel \
        glib2-devel \
        lcms2-devel \
        libexif-devel \
        libgsf-devel \
        libjpeg-turbo-devel \
        libpng-devel \
        libtiff-devel \
        orc-devel 

RUN wget $VIPS_URL/v$VIPS_VERSION/vips-$VIPS_VERSION.tar.gz \
        && tar xzf vips-$VIPS_VERSION.tar.gz \
        && cd vips-$VIPS_VERSION \
        && ./configure --without-heif \
        && make V=0 \
        && make install

# test ruby-vips
RUN gem install ruby-vips \
        && ruby -e 'require "vips"; puts "success!"'
```

(though I've disabled libheif in libvips configure -- try building with the head of the 8.10 branch)

**Comment by estepnv** [comment]
@jcupitt thanks for your reply. however, I managed to buildt it using the following setup.
I used pkg-config and it worked well.

```docker
ROM lambci/lambda:build-ruby2.7

WORKDIR /build

ENV WORKDIR="/build"
ENV INSTALLDIR="/opt"
ENV VIPS_VERSION="8.10.1"

RUN curl -L https://github.com/strukturag/libheif/releases/download/v1.9.1/libheif-1.9.1.tar.gz | tar xz
RUN curl -L https://github.com/libvips/libvips/releases/download/v8.10.1/vips-8.10.1.tar.gz | tar xz

RUN yum install -y gtk-doc
RUN yum install -y gobject-introspection \
  gobject-introspection-devel \
  expat-devel \
  libjpeg-turbo-devel \
  libtiff-devel \
  libpng-devel \
  lcms2-devel \
  mercurial \
  orc-devel \
  glib2-devel

RUN curl http://mirror.centos.org/centos/8/PowerTools/x86_64/os/Packages/nasm-2.13.03-2.el8.x86_64.rpm --output nasm.rpm && yum install -y nasm.rpm

# compiling libx265 dependency for libheif
RUN hg clone http://hg.videolan.org/x265
RUN cd ./x265/build/linux && \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig cmake -G "Unix Makefiles" ../../source && \
  make -j6 && \
  make install && \
  ldconfig

# compiling libde265 dependency for libheif
RUN curl -L https://github.com/strukturag/libde265/releases/download/v1.0.7/libde265-1.0.7.tar.gz | tar xz
RUN cd libde265-1.0.7 && \
  ./autogen.sh && \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig ./configure --disable-sherlock265 && \
  make -j6 && \
  make install

# compiling libheif dependency for libvips
RUN cd libheif-1.9.1 && \
  ./autogen.sh && \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig ./configure && \
  make -j6 && \
  make install

# compling libvips
RUN cd vips-8.10.1 && \
  PKG_CONFIG_PATH=/usr/local/lib/pkgconfig ./configure

RUN cd vips-8.10.1 && \
  make -j6 && \
  make install && \
  echo /opt/lib > /etc/ld.so.conf.d/libvips.conf && \
  echo /usr/lib64 > /etc/ld.so.conf.d/libvips.conf && \
  ldconfig

# assemebling

RUN mkdir -p share/lib && \
  cp -a /usr/local/lib/libvips.so* $WORKDIR/share/lib/ && \
  cp -a /usr/local/lib/libheif.so* $WORKDIR/share/lib/ && \
  cp -a /usr/local/lib/libx265.so* $WORKDIR/share/lib/ && \
  cp -a /usr/local/lib/libde265.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/liborc-0.4.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libpng*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libjbig*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libblkid*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libuuid*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/liblcms2*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libmount*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libgio*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libjpeg*.so* $WORKDIR/share/lib/ && \
  cp -a /usr/lib64/libtiff*.so* $WORKDIR/share/lib/

# Create sym links for ruby-ffi gem's `glib_libname` and `gobject_libname` to work.
RUN cd ./share/lib/ && \
  ln -s /usr/lib64/libglib-2.0.so.0 libglib-2.0.so && \
  ln -s /usr/lib64/libglib-2.0.so.0 glib-2.0.so.0 && \
  ln -s /usr/lib64/libgobject-2.0.so.0 libgobject-2.0.so && \
  ln -s /usr/lib64/libgobject-2.0.so.0 gobject-2.0.so.0 && \
  ln -s /usr/lib64/libgmodule-2.0.so.0 libgmodule-2.0.so.0 && \
  ln -s /usr/lib64/libexpat.so.1 libexpat.so.1


RUN cd ./share && \
  zip --symlinks -r libvips.zip .
```

**Comment by jcupitt** [comment]
Should be fixed in 8.10.2 https://github.com/libvips/libvips/releases/tag/v8.10.2
